src/InjectDmaTransfer.cpp

```diff
@@ -0,0 +1,417 @@
+                let_body = let_body.as<LetStmt>()->body;
+            }
+            is_body_a_single_for_loop = let_body.as<For>() != nullptr;
+        }
+        loop_vars.push_back({op->name, op->min, op->extent, is_body_a_single_for_loop});
+        Stmt mutated = IRMutator::visit(op);
+        loop_vars.pop_back();
+        if (loops_to_be_removed.count(op->name) > 0) {
+            loops_to_be_removed.erase(op->name);
+            return mutated.as<For>()->body;
+        }
+        return mutated;
+    }
+
+    Stmt visit(const LetStmt *op) override {
+        // TODO: Not really correct, but probably want to skip lets which
+        // don't depend on loop vars.
+        if (loop_vars.empty()) {
+            return IRMutator::visit(op);
+        }
+        containing_lets[op->name] = op->value;
+
+        Stmt stmt;
+        Stmt body = mutate(op->body);
+        if (body.same_as(op->body)) {
+            stmt = op;
+        } else {
+            stmt = LetStmt::make(op->name, op->value, body);
+        }
+
+        containing_lets.erase(op->name);
+        return stmt;
+    }
+
+    Stmt visit(const Store *op) override {
+        if (op->name != producer_name) {
+            return IRMutator::visit(op);
+        }
+
+        // Check if the destination is an output buffer in which case we can
+        // do a wait for completion later.
+        is_output_dma = op->param.defined();
+        debug(3) << "InjectDmaTransfer::store " << op->name << "\n";
+        debug(3) << loop_vars.size() << "\n";
+
+        const Load *maybe_load = op->value.as<Load>();
+        if (const Call *maybe_call = op->value.as<Call>()) {
+            if (maybe_call->is_intrinsic(Call::IntrinsicOp::strict_float)) {
+                maybe_load = maybe_call->args[0].as<Load>();
+            }
+        }
+        // Has to be a direct load-to-store for now.
+        user_assert(maybe_load) << "Only direct load-to-stores are supported in dma()";
+
+        debug(3) << "InjectDmaTransfer::" << op->name << " " << maybe_load->name << "\n";
+        debug(3) << op->index << "\n";
+        debug(3) << maybe_load->index << "\n";
+
+        // Substitute in lets into indices of load and store to simplify a further
+        // analysis.
+        Expr op_index = op->index;
+        op_index = substitute_in_all_lets(op_index);
+        op_index = substitute(containing_lets, op_index);
+
+        Expr value_index = maybe_load->index;
+        value_index = substitute_in_all_lets(value_index);
+        value_index = substitute(containing_lets, value_index);
+
+        // A vector to hold DMA extents.
+        std::vector<Expr> dma_extents;
+
+        vector<Expr> store_strides;
+        vector<Expr> value_strides;
+
+        // Compute strides for each of the loop vars.
+        for (const auto &v : loop_vars) {
+            Scope<Expr> local_scope;
+            local_scope.push(v.name, 1);
+            Expr is_linear_store = is_linear(op_index, local_scope);
```

Comment:
op_index and value_index could be combinatorially large due to the substitution above, but is_linear just visits it recursively.